File: original_app/GregoireConverter.gabc/GabcExporter.cs
using System;
using GregoireConverter.GRG;
using System.Resources;
using System.IO;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Runtime.Remoting.Messaging;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Linq.Expressions;
using System.CodeDom.Compiler;
using System.Runtime.Serialization;
using System.Security.Cryptography.X509Certificates;
using System.ComponentModel;

namespace GregoireConverter.gabc
{
	public class GabcExporter : IGRGExporter
	{
		public bool CreatesSingleFile => false;
		public static readonly string GabcTemplateTex = "GregoireConverter.gabc.Resources.gabcTemplate.tex";
		public static readonly string BaseFileNameMarker = "<@!BASE_FILE_NAME!@>";
		public static readonly string DocumentFontSizeMarker = "<@!DOCUMENT_FONT_SIZE!@>";
		public static readonly string OrientationMarker = "<@!ORIENTATION!@>";
		public static readonly string StaffWidthMarker = "<@!STAFF_WIDTH!@>";

		public GabcExporter ()
		{
		}

		public int Save(GRG2 gregoireDocument, string baseDir, string baseFileName, bool overwrite = false)
		{
			//	Check arguments
			if (gregoireDocument == null)
				throw new ArgumentNullException("gregoireDocument", "Argument cannot be null!");
			if (string.IsNullOrWhiteSpace(baseDir))
				throw new ArgumentNullException("baseDir", "Argument cannot be null!");
			if (string.IsNullOrWhiteSpace(baseFileName))
				throw new ArgumentNullException("baseFileName", "Argument cannot be null!");

			//	Ensure that the destination directory is in place
			var dir = new DirectoryInfo(baseDir);
			if (!dir.Exists)
			{
				dir.Create();
			}

			//	Set files names
			var outputTexFile = baseFileName + ".tex";
			var outputGabcFile = baseFileName + ".gabc";

			//	Read TeX template file and save result TeX file
			var assembly = Assembly.GetExecutingAssembly();

			using (var stream = assembly.GetManifestResourceStream(GabcTemplateTex))
			using (var reader = new StreamReader(stream))
			{
				var texTemplate = reader.ReadToEnd();
				using (var outTex = new StreamWriter(Path.Combine(baseDir, outputTexFile)))
				{
					outTex.Write(FillTexTemplate(gregoireDocument, texTemplate, baseFileName));
				}
			}

			//	Generate gabc output file
			var header = new GabcHeader(baseFileName).AddInitial(gregoireDocument.Documents.First());
			var content = GenerateGabcPart(gregoireDocument.Documents.First());

			using (var outGabc = new StreamWriter(Path.Combine(baseDir, outputGabcFile)))
			{
				outGabc.Write(header.ToString());
				outGabc.Write(content.ToString());
			}

			return 0;
		}

		public string FillTexTemplate(GRG2 gregoire, string template, string baseFileName)
		{
			var result = template.Replace(BaseFileNameMarker, baseFileName);

			//	Document properties
			var document = gregoire.Documents.First();
			result = result.Replace(DocumentFontSizeMarker, string.Format("{0:D}", document.FontSize));

			var orientation = "portrait";
			var maxWidth = (int)(1.1*document.Staffs.Max((s) => s.Width)); // gabc staffs are about 10% longer then Gregoire's
			if (maxWidth > 190) // 190mm width staff won't fit in portrait mode
			{
				orientation = "landscape";
			}
			result = result.Replace(OrientationMarker, orientation).Replace(StaffWidthMarker, maxWidth.ToString());

			return result;
		}

		public string GenerateGabcPart(GRG2Document document)
		{
			var result = new StringBuilder();
			GRG2NeumeForGabc lastClef = null;

			for(int i = 0; i < document.Staffs.Count; i++)
			{
				ConvertStaff(document.Staffs[i], i, ref result, ref lastClef);
			}

			return result.ToString();
		}

		public void ConvertStaff(GRG2Staff staff, int staffIdx, ref StringBuilder result, ref GRG2NeumeForGabc lastClef)
		{
			var initialLetter = (staff.Initial != null)? staff.Initial.InitialCaption : string.Empty;

			var textBuf = string.Empty;
			var neumeBuf = string.Empty;

			var neumes = staff.Neumes.OrderBy(n => n.PositionY).OrderBy(n => n.PositionX)
				.Select((n) => new GRG2NeumeForGabc(n)).ToList();

			//	Translate neumes
			for (var it = 0; it < neumes.Count; it++)
			{
				var currentNeume = neumes[it];

				var separator = (it > 0)? GetSeparator(neumes[it - 1], currentNeume) : null;
				//	Captions are not mergeable, so if this neume has next caption, buffers must be empty
				if ((!string.IsNullOrWhiteSpace(textBuf) && !string.IsNullOrWhiteSpace(currentNeume.Caption)))
				{
					FlushBuffers(ref textBuf, ref neumeBuf, ref result);
				}
				//	If the distance is too big to merge those neumes, then also flush.
				else if (separator == null)
				{
					FlushBuffers(ref textBuf, ref neumeBuf, ref result);
				}

				//	Check if this is clef
				if (currentNeume.IsClef)
				{
					if (lastClef != null
						&& lastClef.GetClefIndicator(false)
						.Equals(currentNeume.GetClefIndicator(false)))
					{
						// This is clef and it is the same as the previous one, so it should be skipped.
						continue;
					}

					FlushBuffers(ref textBuf, ref neumeBuf, ref result);

					//	Check if next symbol is bemol. If yes, then combine clef with bemol.
					if (neumes[it + 1].IsBemol
						&& (neumes[it+1].PositionY == currentNeume.PositionY + 6))
					{
						it++;
						neumeBuf += currentNeume.GetClefIndicator(true);
					}
					else
					{
						neumeBuf += currentNeume.GetClefIndicator(false);
					}
					lastClef = currentNeume;

					// Clef should be the first sign of the file
					FlushBuffers(ref textBuf, ref neumeBuf, ref result);
					continue;
				}

				// Custos at the end is added automatically
				if (currentNeume.IsCustos && it == neumes.Count - 1)
				{
					continue;
				}

				// Vertical lines are treated a bit differently;
				if (currentNeume.IsDivisio)
				{
					FlushBuffers(ref textBuf, ref neumeBuf, ref result);
					textBuf += (!string.IsNullOrWhiteSpace(currentNeume.Caption))? currentNeume.Caption : " ";
					try
					{
						neumeBuf = currentNeume.TranslateNeume();
					}
					catch (ArgumentOutOfRangeException ex)
					{
						Logger.LogError(ex);
						Logger.LogWarning("Skipping this neume");
					}

					if (!string.IsNullOrWhiteSpace(neumeBuf))
					{
						FlushBuffers(ref textBuf, ref neumeBuf, ref result);
						textBuf += " ";
					}
					continue;
				}

				//	Check if there is some text for this neume
				if (!string.IsNullOrWhiteSpace(currentNeume.Caption))
				{
					// Initial have to be first letter of text
					if (!string.IsNullOrWhiteSpace(initialLetter))
					{
						textBuf += initialLetter;
						initialLetter = string.Empty;
					}

					// Special case - star in Caption. Stars should be bound to divisio in gabc
					var match = Regex.Match(currentNeume.Caption, @"([*\ ]+)$");
					if (match.Success && neumes[it + 1].IsDivisio)
					{
						currentNeume.Caption = currentNeume.Caption
							.Remove(currentNeume.Caption.Length - match.Groups[1].Length);

						neumes[it + 1].Caption = " *" + neumes[it + 1].Caption;
					}

					textBuf += currentNeume.Caption;
				}

				try
				{
					var neume = currentNeume.TranslateNeume();
					if (!string.IsNullOrWhiteSpace(neume))
						neumeBuf += separator + neume;
				}
				catch (ArgumentOutOfRangeException ex)
				{
					Logger.LogError(ex);
					Logger.LogWarning("Skipping this neume!");
				}
			}

			FlushBuffers(ref textBuf, ref neumeBuf, ref result);

			//	End the line end set justification flag
			result.AppendLine((staff.Justify > 0)? "(z)\n" : "(Z)\n");
		}

		private void FlushBuffers(ref string textBuf, ref string neumeBuf, ref StringBuilder sb)
		{
			if (!(string.IsNullOrWhiteSpace(textBuf) && string.IsNullOrWhiteSpace(neumeBuf))) 
			{
				sb.AppendFormat("{0}({1})", textBuf.TrimEnd(), neumeBuf);
				
				textBuf = (textBuf.EndsWith(" "))? " " : string.Empty;
				neumeBuf = string.Empty;
			}
		}

		private string GetSeparator(GRG2NeumeForGabc previous, GRG2NeumeForGabc current)
		{
			var distance = current.PositionX - (previous.PositionX + previous.SizeX);

			if (distance < 1)
				return "!";
			if (distance < 3)
				return "/";
			if (distance < 5)
				return "//";
			if (distance < 7)
				return " ";

			return null;
		}
	}
}

File: web/grg2gabc/src/app/core/gabc/gabc-exporter.ts
import { Grg2 } from '../grg2/grg2.model';
import { Grg2Document } from '../grg2/grg2-document.model';
import { Grg2Staff } from '../grg2/grg2-staff.model';
import { GabcHeader } from './gabc-header';
import { Grg2NeumeForGabc, UnmappedNeume } from './grg2-neume-for-gabc';

export interface ConversionResult {
  gabc: string;
  warnings: UnmappedNeume[];
}

const UNMAPPED_PLACEHOLDER = '(?)';

export class GabcExporter {
  private warnings: UnmappedNeume[] = [];

  convert(grg: Grg2, fileName: string): ConversionResult {
    this.warnings = [];
    const doc = grg.documents[0];
    if (!doc) throw new Error('GRG2 file contains no documents');

    const header = new GabcHeader(fileName).addInitial(doc);
    const body = this.generateGabcPart(doc);

    return { gabc: header.toString() + body, warnings: this.warnings };
  }

  private generateGabcPart(document: Grg2Document): string {
    let result = '';
    let lastClef: Grg2NeumeForGabc | null = null;

    for (let i = 0; i < document.staffs.length; i++) {
      result += this.convertStaff(document.staffs[i], i, lastClef, (clef) => { lastClef = clef; });
    }
    return result;
  }

  private convertStaff(
    staff: Grg2Staff,
    _staffIdx: number,
    lastClef: Grg2NeumeForGabc | null,
    setLastClef: (c: Grg2NeumeForGabc) => void,
  ): string {
    let result = '';
    let initialLetter = staff.initial?.initialCaption ?? '';
    let textBuf = '';
    let neumeBuf = '';

    const neumes = [...staff.neumes]
      .sort((a, b) => a.positionY - b.positionY)
      .sort((a, b) => a.positionX - b.positionX)
      .map(n => new Grg2NeumeForGabc(n));

    const flush = (): void => {
      if (textBuf.trim() || neumeBuf.trim()) {
        result += `${textBuf.trimEnd()}(${neumeBuf})`;
        textBuf = textBuf.endsWith(' ') ? ' ' : '';
        neumeBuf = '';
      }
    };

    for (let it = 0; it < neumes.length; it++) {
      const cur = neumes[it];
      const separator = it > 0 ? this.getSeparator(neumes[it - 1], cur) : null;

      if ((textBuf.trim() && cur.caption.trim()) || separator === null) {
        flush();
      }

      if (cur.isClef) {
        if (lastClef && lastClef.getClefIndicator(false) === cur.getClefIndicator(false)) continue;
        flush();
        const nextNeume = neumes[it + 1];
        if (nextNeume?.isBemol && nextNeume.positionY === cur.positionY + 6) {
          it++;
          neumeBuf += cur.getClefIndicator(true);
        } else {
          neumeBuf += cur.getClefIndicator(false);
        }
        setLastClef(cur);
        flush();
        continue;
      }

      if (cur.isCustos && it === neumes.length - 1) continue;

      if (cur.isDivisio) {
        flush();
        textBuf += cur.caption.trim() ? cur.caption : ' ';
        const translated = this.translateNeume(cur);
        neumeBuf = translated;
        if (neumeBuf.trim()) flush();
        textBuf += ' ';
        continue;
      }

      if (cur.caption.trim()) {
        if (initialLetter) {
          textBuf += initialLetter;
          initialLetter = '';
        }

        // Move trailing star to the following divisio
        const starMatch = cur.caption.match(/([* ]+)$/);
        if (starMatch && neumes[it + 1]?.isDivisio) {
          cur.caption = cur.caption.slice(0, cur.caption.length - starMatch[1].length);
          neumes[it + 1].caption = ' *' + neumes[it + 1].caption;
        }

        textBuf += cur.caption;
      }

      const translated = this.translateNeume(cur);
      if (translated) neumeBuf += (separator ?? '') + translated;
    }

    flush();
    result += (staff.justify > 0) ? '(z)\n\n' : '(Z)\n\n';
    return result;
  }

  private translateNeume(neume: Grg2NeumeForGabc): string {
    if (!neume.format) {
      this.warnings.push({ id: neume.id, positionX: neume.positionX, positionY: neume.positionY, caption: neume.caption });
      return UNMAPPED_PLACEHOLDER;
    }
    try {
      return neume.translateNeume();
    } catch {
      this.warnings.push({ id: neume.id, positionX: neume.positionX, positionY: neume.positionY, caption: neume.caption });
      return UNMAPPED_PLACEHOLDER;
    }
  }

  private getSeparator(prev: Grg2NeumeForGabc, cur: Grg2NeumeForGabc): string | null {
    const distance = cur.positionX - (prev.positionX + prev.sizeX);
    if (distance < 1)  return '!';
    if (distance < 3)  return '/';
    if (distance < 5)  return '//';
    if (distance < 7)  return ' ';
    return null;
  }
}
